fw/mmap_buffer.c

```diff
@@ -0,0 +1,303 @@
+		return -EAGAIN;
+
+	vma->vm_ops = &dev_vm_ops;
+	(void)dev_vm_ops;
+
+	if (size == buf_size)
+		atomic_set(&buf->is_ready, 1);
+
+	return 0;
+
+#undef NTH_ONLINE_CPU
+}
+
+static void dev_file_vm_close(struct vm_area_struct *vma)
+{
+	TfwMmapBufferHolder *holder = vma->vm_file->private_data;
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+
+	atomic_set(&buf->is_ready, 0);
+}
+
+TfwMmapBufferHolder *
+tfw_mmap_buffer_create(const char *filename, unsigned int size)
+{
+	TfwMmapBufferHolder *holder;
+	unsigned int order;
+	int cpu;
+
+	if (size < TFW_MMAP_BUFFER_MIN_SIZE
+		|| size > TFW_MMAP_BUFFER_MAX_SIZE
+		|| !is_power_of_2(size))
+		return NULL;
+
+	if (filename && strlen(filename) >= TFW_MMAP_BUFFER_MAX_NAME_LEN - 1)
+		return NULL;
+
+	holder = kmalloc(sizeof(TfwMmapBufferHolder) +
+					 sizeof(struct page *) * num_online_cpus(),
+					 GFP_KERNEL);
+	if (!holder)
+		return NULL;
+
+	order = get_order(size);
+
+	holder->dev_major = -1;
+	holder->buf = __alloc_percpu_gfp(sizeof(TfwMmapBuffer *),
+									 sizeof(u64), GFP_KERNEL);
+	atomic_set(&holder->dev_is_opened, 0);
+	atomic_set(&holder->is_freeing, 0);
+
+	for_each_online_cpu(cpu) {
+		TfwMmapBuffer *buf, **bufp;
+
+		holder->pg[cpu] = alloc_pages_node(cpu_to_node(cpu), GFP_KERNEL, order);
+		if (holder->pg[cpu] == NULL)
+			goto err;
+
+		buf = (TfwMmapBuffer *)page_address(holder->pg[cpu]);
+		buf->size = TFW_MMAP_BUFFER_DATA_SIZE(size);
+		buf->head = 0;
+		buf->tail = 0;
+		buf->cpu = cpu;
+		bufp = per_cpu_ptr(holder->buf, cpu);
+		*bufp = buf;
+		atomic_set(&buf->is_ready, 0);
+	}
+
+	if (filename) { /* do not create the file in unit tests */
+		holder->dev_major = register_chrdev(0, filename, &dev_fops);
+		if (holder->dev_major < 0) {
+			T_WARN("Registering char device failed for %s\n", filename);
+			goto err;
+		}
+
+		holder->dev_class = class_create(THIS_MODULE, filename);
+		device_create(holder->dev_class, NULL,
+					  MKDEV(holder->dev_major, 0), NULL, filename);
+		strscpy(holder->dev_name, filename, sizeof(holder->dev_name));
+		holders[holders_cnt++] = holder;
+	}
+
+	return holder;
+
+err:
+	tfw_mmap_buffer_free(holder);
+
+	return NULL;
+}
+
+void
+tfw_mmap_buffer_free(TfwMmapBufferHolder *holder)
+{
+	int cpu;
+
+	if (!holder)
+		return;
+
+	atomic_set(&holder->is_freeing, 1);
+
+	for_each_online_cpu(cpu) {
+		TfwMmapBuffer *buf = *per_cpu_ptr(holder->buf, cpu);
+		/* Notify user space that it have to close the file */
+		atomic_set(&buf->is_ready, 0);
+	}
+
+	/* Wait till user space closes the file */
+	while (atomic_read(&holder->dev_is_opened))
+		schedule();
+
+	for_each_online_cpu(cpu) {
+		TfwMmapBuffer *buf = *per_cpu_ptr(holder->buf, cpu);
+
+		if (holder->pg[cpu]) {
+			__free_pages(holder->pg[cpu],
+						 get_order(TFW_MMAP_BUFFER_FULL_SIZE(buf->size)));
+			holder->pg[cpu] = NULL;
+		}
+	}
+
+	if (holder->dev_major > 0) {
+		device_destroy(holder->dev_class, MKDEV(holder->dev_major, 0));
```

Comment:
It seems that deinitialization should be more complicated, because device can be NULL and class can be NULL

Reply:
Fixed.